### PR TITLE
fix smooth scrolling on macOS

### LINKED ISSUE #6068: macOS: lighttable scrolling nearly unusable and preview very slow since 3.2.1
**Environment**

- macOS 10.15.6
- Macbook Pro 16, i9-9880H, 16GB, SSD
- openCL devices: 'Intel(R) UHD Graphics 630' (device 0), 'AMD Radeon Pro 5500M Compute Engine' (device 1)
- Darktable 3.2.1 / 3.0.2 directly downloaded as .DMG from Darktable's official GitHub releases
- about 800 RAW images (.NEF) of a Nikon D750

**Steps to reproduce**

1. Remove any previous /Applications/darktable.app
2. Clean config/cache (delete ~/.config/darktable and ~/.cache/darktable)
3. Do a fresh install of 3.2.1 / 3.0.2 without changing any configuration. _Note:_ Due to default setting, OpenCL is not activated
5. Import images
6. Scroll through images e.g. in a 5 images per row grid (by using trackpad or mouse wheel)
7. Open a preview (selecting one image and pressing "w")

**Current behaviour (as in 3.2.1)**

- images get imported without showing any thumbnails during import process
- thumbnails start popping up (being displayed one after the other) only after import has finished
- scrolling is not smooth and in an uncontrolled manner (a swipe brings you somewhere much further within the collection; 2-3 swipes can bring you at the end of the collection; two down and two up will not bring you to the same images/row)
- during scrolling thumbnails are popping in, nearly one after the other as if they are being loaded from a very slow disk
- preview takes 2-3 seconds to display

**Expected behaviour (as in 3.0.2)**

- images get imported whilst thumbnails are shown instantly as they are being added without any delay
- scrolling through thumbnails is smooth and in a controlled manner (a swipe brings you to the next images; you can do two swipes down and then two up again to get to the same images/row)
- during scrolling thumbnails are displayed instantly, no delay 
- preview is shown instantly without any delay

**Further notes**

- activating OpenCL => "multiple GPUs" did not bring any significant improvements if any at all
- tried e.g. opencl_device_priority=!Intel(R) UHD Graphics 630,*/!AMD Radeon Pro 5500M Compute Engine,*/AMD Radeon Pro 5500M Compute Engine,*/AMD Radeon Pro 5500M Compute Engine,*
- no difference when scrolling with mouse or trackpad
- changing theme does not have any effect on the described behaviour
- same workflow on a rather old, slow Linux machine is blazing fast with 3.2.1

**Workaround**

- scroll with cursor keys
- accept slowness and have coffee
- get a Linux machine

### PR DESCRIPTION
**Pull request by lhietal**
Fixing issue on macOS with smooth scrolling events. Fixes #6068 scrolling issue by scaling the input from _event->delta_x_ and _event->delta_y_.

The scaling amount feels ok on my system. It could be changed to get faster scrolling. Besides lighttable this also affects some modules. I tested module resize, retouch, histogram, zone system and bauhaus comboboxes. They seem to work correctly.

There might be better ways to handle trackpads. https://github.com/darktable-org/darktable/issues/6068#issuecomment-751350041

### CODE REVIEW

**Comment by parafin** [comment]
Can you tell where the formula comes from or how it was created?

**Comment by lhietal** [comment]
I was just trying to fit input to the expectation that scrolling happens when 1 is returned. I have no spesific information on what values from _event->delta_x_ represent. 

It is a general mapping between ranges. https://stackoverflow.com/questions/5731863/mapping-a-numeric-range-onto-another

**Comment by parafin** [comment]
I'm asking, because https://github.com/Beep6581/RawTherapee/issues/5599 issue you linked in #6068 got resolved differently: https://github.com/Beep6581/RawTherapee/pull/5607
They are basically ignoring the delta_x value altogether and just use 1 instead as value. As the source they have GTK own code for scrollbars. So I'm wondering which approach is better. Have you tried both options?

**Comment by lhietal** [comment]
I didn't underestand the RawTherapee solution completely but I think they use _event->delta_x_ somehow differently.

The way I underestand _dt_gui_get_scroll_unit_deltas()_ is designed to work is that it sets *delta_x  to 1 or -1 when multiple smaller values from _event->delta_x_ accumulate to 1 or -1. Thumbtable is then moved by one row. On macOS _event->delta_x_ is a lot larger than 1. So scrolling is too fast.
Setting acc_x = 1 still causes tumbtable to move every time the function is called. Which is several times per second.

**Comment by MStraeten** [comment]
this fix doesn't improve the scroll behaviour in lighttable filemanager on a magic mouse or trackpad. It's very hard to scroll just one row - one little movement and you've jumped over 20 rows ...

**Comment by MStraeten** [comment]
after adding some debug messages in the code indicates the *get_scroll* functions are never used for the scrollbars of the filemanager/thumbtable in lighttable

**Comment by lhietal** [comment]
_dt_gui_get_scroll_unit_delta()_ is called in src/dtgtk/thumbtable.c on line 893. Am I underestanding the functionality of __event_scroll()_ incorrectly?

**Comment by MStraeten** [comment]
gtk.c line 695:
```
gboolean dt_gui_get_scroll_unit_delta(const GdkEventScroll *event, int *delta)
{
  int delta_x, delta_y;
   fprintf(stderr, "dt_gui_get_scroll_unit_delta\n");
  if(dt_gui_get_scroll_unit_deltas(event, &delta_x, &delta_y))
  {
    *delta = delta_x + delta_y;
    return TRUE;
  }
  return FALSE;
}
```
The debug message isn't printed when having scrollbars for central view in lighttable and darkroom activated in preferences and scrolling while having them in focus. 
If scrolling is done inside the central view it's called (but thats doesn't affect the scrollbars) - and in this case the pr works fine.

File: src/gui/gtk.c [hunk 1]
```diff
@@ -654,8 +654,13 @@ gboolean dt_gui_get_scroll_unit_deltas(const GdkEventScroll *event, int *delta_x
       // accumulate trackpad/touch scrolls until they make a unit
       // scroll, and only then tell caller that there is a scroll to
       // handle
+#ifdef GDK_WINDOWING_QUARTZ // on macOS deltas need to be scaled
+      acc_x += (2.0 / 300.0) * (event->delta_x + 150) - 1;
```
[the remark this answers is not in the document]
Comment (lhietal) [reply src/gui/gtk.c:658]: No. The range should be symmetrical around 0 so simplifying works.

**Comment by lhietal** [comment]
> The debug message isn't printed when having scrollbars for central view in lighttable and darkroom activated in preferences and scrolling while having them in focus.

In this case it seems that scroll distance comes directly from GtkScrollbar widget in _scrollbar_changed()_ (gtk.c line 871). Scroll events are not handled directly so same fix is not possible.

File: src/gui/gtk.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (MStraeten) [reply src/gui/gtk.c:658]: after using this a couple of days if find the divisor 150 makes the ui very laggy - i found a value of 50 being more responsive but not too fast (at least for my old magic mouse)

[the remark this answers is not in the document]
Comment (lhietal) [reply src/gui/gtk.c:658]: You are right. It is better to use the smaller value. It is still possible to scroll by one row.

**Comment by MStraeten** [comment]
to get this stuff done:
@lhieta please change the code as suggested, so @parafin can check and merge.
tweaking the behaviour of the scrollbar widget is a different story ...

**Comment by parafin** [comment]
I would still like to know how well rawtherapee solution (copied from GTK) work. Which is it seems just replacing `event->delta_x / 50` with `1` or some other constant.
But if current version of PR is confirmed to make things better I'm fine with merging it.

**Comment by jonastr** [comment]
@parafin I'm not sure I understand your question, so maybe my answer is not what you're looking for: There are no scrolling issues in RawTherapee on MacOS. 
Or is your question rather to try and copy the same bits to see if it works with Darktable as well?

**Comment by MStraeten** [comment]
It quite easy to see the effect when scrolling the thumbtable in lighttable mode with magic mouse or trackpad since the slider widgets aren’t affected. 
using the slider a little swipe scrolls over 20 lines. Just changing the angle the finger is touching the mouse can cause scrolling over a couple of rows.
If the mousepointer is on the thumbtable it’s now possible to scroll by row as it is done when using a real mousewheel.
So even it doesn‘t affect the scrollbarwidget its a real improvement.

**Comment by parafin** [comment]
> @parafin I'm not sure I understand your question, so maybe my answer is not what you're looking for: There are no scrolling issues in RawTherapee on MacOS.
> Or is your question rather to try and copy the same bits to see if it works with Darktable as well?

Yes, there are no issues, because they fixed them, see my previous comment: https://github.com/darktable-org/darktable/pull/7904#issuecomment-761818818

So my questions is why not use an already tested solution?

**Comment by lhietal** [comment]
I haven't spend a lot of time trying to underestend the RawTherapee solution. Simply changing `event->delta_x / 50` to `1` produces similar behavior than before the fix. Changing it to `0.1` makes scrolling slower but doesn't allow scrolling speed to change. So that scrolling is always slow.

**Comment by parafin** [comment]
OK, so let's re-iterate - current PR improves the situation by fixing some scrolling behaviour in darktable, but not all of them (specifically lighttable scrollbar)?

**Comment by MStraeten** [comment]
yes, because the handlers for scroll events in gtk.c aren't triggered if the thumbtable scrollbar widget is used.

**Comment by parafin** [comment]
Merged to master and 3.4 branch

**Pull request merged**